compiler/translator.c

```diff
@@ -94,6 +96,82 @@ int isValidDigit(char *token) {
     return 0;
 }
 
+/**
+ * Checks whether the supplied token is a valid character.
+ * @param token The supplied token
+ * @return the quoted character if it's valid, -1 otherwise
+ */
+char* validChar(char *token)
+{
+    int len = strlen(token);
+    if (len == 1) {
+        // Backslashes should also work, so we escape them
+        if (token[0] == '\\') {
+            return "92";
+        } else if (token[0] == '\'') {
+            return "39";
+        } else {
+            char* newtok = (char *)malloc(3);
+            newtok[0] = '\'';
+            strcpy(&newtok[1], token);
+            newtok[2] = '\'';
+            return newtok;
+        }
+    } else if (len == 2 && token[0] == '\\') {
+        // Escape sequences according to https://en.wikipedia.org/wiki/Escape_sequences_in_C#Table_of_escape_sequences
+        // We just return their ASCII number as a string as that can also be used to print the string
+        // This is mostly done because when writing '\n' in asm, it prints just '\', which is weird
+        switch (token[1])
+        {
+        case 'b':
+            return "8";
+        case 't':
+            return "9";
+        case 'n':
+            return "10";
+        case 'v':
+            return "11";
+        case 'f':
+            return "12";
+        case 'r':
+            return "13";
+        case '"':
+            return "34";
+        case '\'':
+            return "39";
+        case '?':
+            return "63";
+        case '\\':
+            return "92";
+        case 's':
+            // This usually isn't a valid escape character, but we treat it as a space
+            return "32";
+        }
+    } else if (strcmp(token, "space") == 0)
+    {
+        return "32";
+    }
+
+    // Invalid
+    return -1;
```

Comment:
This function returns a pointer and not an integer, so returning -1 is not ideal and could lead to a segmentation Fault if not handled correctly. (ah yes, the good'old segFault) It also creates a compiler warning when running the Makefile

My suggestion: Change it to return a NULL-pointer
`return NULL;`

Reply:
Nice, that was exactly what I was looking for :D